underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsActiveSyncProvider.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.hadoop.hdfs.client.HdfsAdmin;
+import org.apache.hadoop.hdfs.inotify.Event;
+import org.apache.hadoop.hdfs.inotify.EventBatch;
+import org.apache.hadoop.hdfs.inotify.MissingEventsException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class HdfsActiveSyncProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(HdfsActiveSyncProvider.class);
+  private final HdfsAdmin mHdfsAdmin;
+  private static final int MAX_ACTIVITY = 10;
+  private static final int MAX_AGE = 10;
+
+  // a map mapping SyncPoints to a set of files that have been changed under that syncPoint
+  private Map<String, Set<AlluxioURI>> mChangedFiles;
+  // Use an integer to indicate the activity level of the sync point
+  private Map<String, Integer> mActivity;
+  private Map<String, Integer> mAge;
+  private DFSInotifyEventInputStream mEventStream;
+
+  public HdfsActiveSyncProvider(HdfsAdmin hdfsAdmin) throws IOException {
+    mHdfsAdmin = hdfsAdmin;
+    mChangedFiles = new ConcurrentHashMap<>();
+    mActivity = new ConcurrentHashMap<>();
+    mAge = new ConcurrentHashMap<>();
+    mEventStream = mHdfsAdmin.getInotifyEventStream();
+  }
+
+  private void initNextWindow() {
+    for (String syncPoint : mActivity.keySet()) {
+      mActivity.put(syncPoint, new Integer(mActivity.get(syncPoint).intValue() / 10));
+      mAge.put(syncPoint, new Integer(mAge.get(syncPoint).intValue() + 1));
+    }
+  }
+
+  private void addFile(String syncPoint, String filePath) {
+    LOG.info("add file for syncPoint {}, at path {}", syncPoint, filePath);
+    AlluxioURI syncPointUri = new AlluxioURI(syncPoint);
+    if (!mChangedFiles.containsKey(syncPoint)){
+      mChangedFiles.put(syncPoint, new ConcurrentHashSet<>());
+      mActivity.put(syncPoint, new Integer(0));
+      mAge.put(syncPoint, new Integer(0));
+    }
+    mChangedFiles.get(syncPoint).add(
+        new AlluxioURI(syncPointUri.getScheme(), syncPointUri.getAuthority(), filePath));
+    mActivity.put(syncPoint, new Integer(mActivity.get(syncPoint).intValue() + 1));
+  }
+
+  private void clearFile(String syncPoint){
+    LOG.info("clear file for syncPoint {}, ", syncPoint );
+    mChangedFiles.remove(syncPoint);
+    mActivity.remove(syncPoint);
+    mAge.remove(syncPoint);
+  }
+
+  private boolean isEventModification(Event inotifyEvent) {
+    List<Event.EventType> eventList = Arrays.asList(new Event.EventType[]{Event.EventType.APPEND,
+        Event.EventType.CREATE, Event.EventType.METADATA, Event.EventType.RENAME,
+        Event.EventType.TRUNCATE, Event.EventType.UNLINK});
+    return eventList.contains(inotifyEvent.getEventType());
+  }
+
+  public boolean processEvent(Event event, List<AlluxioURI> syncPointList) {
+    boolean isModified = isEventModification(event);
```

Comment:
why do we need this boolean, if we are switching on all the modification events below?